### ISSUE TITLE
Failed to create etcd cluster client

### ISSUE THREAD

**Issue by matthewmrichter** [Quentin-M/etcd-cloud-operator#8]
I'm building a pilot ECO cluster. However, I'm getting "failed to create etcd cluster client - error=dial tcp <local private IP>:2379: getsockopt: connection refused". 

Host: CentOS 7, Docker version 18.04.0-ce, build 3d479c0

eco.yaml: 
```eco:
  # The interval between each cluster verification by the operator.
  check-interval: 15s
  # The time after which, an unhealthy member will be removed from the cluster.
  unhealthy-member-ttl: 30s
  # Defines whether the operator will attempt to seed a new cluster from a
  # snapshot after the managed cluster has lost quorum.
  auto-disaster-recovery: true
  # Configuration of the etcd instance.
  etcd:
    # The address that clients should use to connect to the etcd cluster (i.e.
    # load balancer public address).
    advertise-address: http://<aws nlb>:2379
    # The directory where the etcd data is stored.
    data-dir: /var/lib/etcd
    # The TLS configuration for clients communication.
    client-transport-security:
      auto-tls: false
      cert-file: /crt/myorg.crt
      key-file: /crt/myorg.pem
      trusted-ca-file: /crt/ca.crt
      client-cert-auth: false
    # The TLS configuration for peers communication.
    peer-transport-security:
      auto-tls: true
      cert-file: /crt/myorg.crt
      key-file: /crt/myorg.key
      trusted-ca-file: /crt/ca.crt
      peer-client-cert-auth: false
    # Defines the maximum amount of data that etcd can store, in bytes, before going into maintenance mode
    backend-quota: 2147483648
  # Configuration of the auto-scaling group provider.
  asg:
    provider: aws
  # Configuration of the snapshot provider.
  snapshot:
    provider: s3
    # The interval between each snapshot.
    interval: 30m
    # The time after which a backup has to be deleted.
    ttl: 24h
    # The bucket where snapshots are stored when using the S3 provider.
    bucket: myorg.eco-test
```

Running with: `docker run -e "AWS_ACCESS_KEY_ID=<access key>" -e "AWS_SECRET_ACCESS_KEY=<secret key>" etcd-cloud-operator`

Getting this output:

```
time="2018-04-18T21:02:13Z" level=info msg="loaded configuration file /etc/eco/eco.yaml"
time="2018-04-18T21:02:19Z" level=warning msg="failed to create etcd cluster client" error="dial tcp <private IP of host>:2379: getsockopt: connection refused"
time="2018-04-18T21:02:19Z" level=warning msg="failed to query <host instance>'s ECO instance" error="Get http://<private IP of host>:2378/status: dial tcp <private IP of host>:2378: getsockopt: connection refused"
time="2018-04-18T21:02:19Z" level=warning msg="failed to query <other asg instance>'s ECO instance" error="Get http://<other asg instance private ip>:2378/status: dial tcp <other asg instance private ip>:2378: getsockopt: connection refused"
time="2018-04-18T21:02:19Z" level=warning msg="failed to query <other asg instance>'s ECO instance" error="Get http://<other asg instance private ip>:2378/status: dial tcp <other asg instance private ip>:2378: getsockopt: connection refused"
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/quentin-m/etcd-cloud-operator/pkg/operator.fetchStatuses(0xc420151470, 0x0, 0xc4204d51c0, 0x3, 0x4, 0x1dd10a0, 0xc4204eed80, 0xc42025e140, 0x14)
	/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/misc.go:120 +0x494
github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).evaluate(0xc420068d00, 0x14668a8, 0xc420068d00)
	/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:134 +0x247
github.com/quentin-m/etcd-cloud-operator/pkg/operator.(*Operator).Run(0xc420068d00)
	/go/src/github.com/quentin-m/etcd-cloud-operator/pkg/operator/operator.go:103 +0x51
main.main()
	/go/src/github.com/quentin-m/etcd-cloud-operator/cmd/operator/operator.go:60 +0x399
```

Any idea what gives? I tried passing log-level debug but no more helpful info on WHY exactly the etcd cluster client failed to create. Also, I've verified selinux, iptables are disabled.

**Comment by matthewmrichter** [comment]
Oh boy. Literally right after I submitted this, I found the issue. Have to run docker with `--net=host`.

**Comment by Quentin-M** [comment]
Hi there,

The connection refused and failed to create the etcd client at the very
beginning of the cluster lifecycle is totally fine.

Good catch with `--net=host`, however, I would hope that ECO would not bail
like so with an index out of range but provide a better message. I will
surely look into it as soon as I can!

Best Regards,
Quentin Machu

On April 18, 2018 at 2:09:14 PM, Matt Richter (notifications@github.com)
wrote:

Oh boy. Literally right after I submitted this, I found the issue. Have to
run docker with --net=host.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/Quentin-M/etcd-cloud-operator/issues/8#issuecomment-382530170>,
or mute the thread
